### PR TITLE
backend/feat/getDownloadInvoiceDataAPI

### PR DESCRIPTION
**Pull request by Anuragini-2506**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [x] This PR contains API breaking changes

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs [hunk 1]
```diff
@@ -114,6 +114,19 @@ findWindowsWithStatus (Id driverId) from to mbStatus limitVal offsetVal =
     (Just limitVal)
     (Just offsetVal)
 
+findWindows :: MonadFlow m => Id Person -> UTCTime -> UTCTime -> m [DriverFee]
+findWindows (Id driverId) from to = do
+  findAllWithOptionsKV
+    [ Se.And
+        [ Se.Is BeamDF.startTime $ Se.GreaterThanOrEq from,
+          Se.Is BeamDF.endTime $ Se.LessThanOrEq to,
+          Se.Is BeamDF.driverId $ Se.Eq driverId
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs:123]: Lets also have an extra filter on `status` where `status != 'INACTIVE'`

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs [hunk 2]
```diff
@@ -170,6 +170,11 @@ type API =
              :<|> "dues"
                :> TokenAuth
                :> Get '[JSON] DDriver.DriverDuesResp
+             :<|> "getDownloadInvoiceData"
```
[the remark this answers is not in the document]
Comment (Anuragini-2506) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs:173]: done

**Review by AshishSaini25 — APPROVED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 3]
```diff
@@ -1824,3 +1793,64 @@ calcExecutionTime transporterConfig' autopayPaymentStage scheduledAt = do
...
+  mapM buildDriverFeeRespEntity driverFees
+  where
+    buildDriverFeeRespEntity DDF.DriverFee {..} = do
+      mbInvoice <- QINV.findLatestByDriverFeeId id
+      return
+        DriverFeeResp
+          { chargesBreakup = mkChargesBreakup govtCharges platformFee.fee platformFee.cgst platformFee.sgst,
+            totalRides = numRides,
+            driverFeeId = id,
+            debitedOn = case mbInvoice of
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1830]: mbInvoice <&> (.updatedAt)

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs [hunk 4]
```diff
@@ -187,6 +187,20 @@ findWindows (Id driverId) from to limitVal offsetVal =
...
     (Just offsetVal)
 
+findWindowsWithoutLimit :: MonadFlow m => Id Person -> UTCTime -> UTCTime -> m [DriverFee]
+findWindowsWithoutLimit (Id driverId) from to = do
+  findAllWithOptionsKV
+    [ Se.And
+        [ Se.Is BeamDF.startTime $ Se.GreaterThanOrEq from,
+          Se.Is BeamDF.endTime $ Se.LessThanOrEq to,
+          Se.Is BeamDF.driverId $ Se.Eq driverId,
+          Se.Is BeamDF.status $ Se.Not $ Se.In [INACTIVE]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs:197]: Just do Eq.

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 5]
```diff
@@ -1826,3 +1795,58 @@ calcExecutionTime transporterConfig' autopayPaymentStage scheduledAt = do
...
+            driverFeeId = id,
+            debitedOn = mbInvoice <&> (.updatedAt),
+            invoiceStatus = mbInvoice <&> (.invoiceStatus),
+            ..
+          }
+
+    mkChargesBreakup _ platformFee cgst sgst =
+      [ DriverPaymentBreakup
+          { component = "Final Platform Fee",
+            amount = platformFee
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1838]: platformFee + cgst + sgst

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 6]
```diff
@@ -1826,3 +1795,58 @@ calcExecutionTime transporterConfig' autopayPaymentStage scheduledAt = do
...
+          },
+        DriverPaymentBreakup
+          { component = "Platform Fee",
+            amount = platformFee - (cgst + sgst)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:1842]: Just platformFee

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs [hunk 7]
```diff
@@ -187,6 +187,20 @@ findWindows (Id driverId) from to limitVal offsetVal =
...
     (Just offsetVal)
 
+findWindowsWithoutLimit :: MonadFlow m => Id Person -> UTCTime -> UTCTime -> m [DriverFee]
+findWindowsWithoutLimit (Id driverId) from to = do
+  findAllWithOptionsKV
+    [ Se.And
+        [ Se.Is BeamDF.startTime $ Se.GreaterThanOrEq from,
+          Se.Is BeamDF.endTime $ Se.LessThanOrEq to,
+          Se.Is BeamDF.driverId $ Se.Eq driverId,
+          Se.Is BeamDF.status $ Se.Not $ Se.Eq INACTIVE
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs:197]: We should not show invoice for ongoing rides too, right?

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs (hunk 7 above)
Comment (Anuragini-2506) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverFee.hs]: I added the condition for ONGOING in the query.

**Review by hkmangla — APPROVED** [review]

**Pull request merged**